server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,318 @@
+            String clusterAlias = entry.getKey();
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+            assert clusterRef.get().getTotalShards() == null : "total shards should not be set on a Cluster before onListShards";
+
+            Integer totalCount = entry.getValue();
+            Integer skippedCount = skippedByClusterAlias.getOrDefault(clusterAlias, 0);
+            TimeValue took = null;
+
+            boolean swapped;
+            do {
+                SearchResponse.Cluster curr = clusterRef.get();
+                SearchResponse.Cluster.Status status = curr.getStatus();
+                assert status == SearchResponse.Cluster.Status.RUNNING : "should have RUNNING status during onListShards but has " + status;
+
+                // if all shards are marked as skipped, the search is done - mark as SUCCESSFUL
+                if (skippedCount != null && skippedCount == totalCount) {
+                    took = new TimeValue(timeProvider.buildTookInMillis());
+                    status = SearchResponse.Cluster.Status.SUCCESSFUL;
+                }
+
+                SearchResponse.Cluster updated = new SearchResponse.Cluster.Builder(curr).setStatus(status)
+                    .setTotalShards(totalCount)
+                    .setSuccessfulShards(skippedCount)
+                    .setSkippedShards(skippedCount)
+                    .setFailedShards(0)
+                    .setTook(took)
+                    .setTimedOut(false)
+                    .build();
+
+                swapped = clusterRef.compareAndSet(curr, updated);
+                assert swapped : "compareAndSet in onListShards should never fail due to race condition";
+            } while (swapped == false);
+        }
+    }
+
+    /**
+     * Executed when a shard returns a query result.
+     *
+     * @param shardIndex  The index of the shard in the list provided by {@link SearchProgressListener#onListShards}.
+     * @param queryResult QuerySearchResult holding the result for a SearchShardTarget
+     */
+    @Override
+    public void onQueryResult(int shardIndex, QuerySearchResult queryResult) {
+        // we only need to update Cluster state here if the search has timed out, since:
+        // 1) this is the only callback that gets search timedOut info and
+        // 2) the onFinalReduce will get all these shards again so the final accounting can be done there
+        // for queries that did not time out
+        if (queryResult.searchTimedOut() && clusters.hasClusterObjects()) {
+            SearchShardTarget shardTarget = queryResult.getSearchShardTarget();
+            String clusterAlias = shardTarget.getClusterAlias();
+            if (clusterAlias == null) {
+                clusterAlias = RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY;
+            }
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+            boolean swapped;
+            do {
```

Comment:
This whole loops could just be:

```java
            clusterRef.updateAndGet(c -> {
                if (c.isTimedOut()) {
                    return c; // cluster has already been marked as timed out on some other shard
                }
                if (c.getStatus() == SearchResponse.Cluster.Status.FAILED || c.getStatus() == SearchResponse.Cluster.Status.SKIPPED) {
                    return c; // safety check to make sure it hasn't hit a terminal FAILED/SKIPPED state where timeouts don't matter
                }
                return new SearchResponse.Cluster.Builder(c).setTimedOut(true).build();
            });
```

I think?

Reply:
Maybe. I find that hard to reason about for correctness, though. I had to read through the updateAndGet method to look at its internals and it's not obvious to me that those `return` statements will work?

This version saves about 5 lines of code, but at the expense of being harder to reason about and now looking different from all the other methods in this class. It would take some time for the code reader to wonder why we do it differently in this method vs. the other.

So I didn't change it yet. Let me know if you want to discuss further.